ni_measurementlink_service/_internal/parameter/serializer.py

```diff
@@ -180,3 +198,44 @@ def _get_missing_parameters(
                 value.type, value.repeated
             )
     return missing_parameters
+
+
+def _deserialize_enum_parameters(
+    parameter_metadata_dict: Dict[int, ParameterMetadata], parameter_by_id: Dict[int, Any]
+):
+    """Converts all enums in `parameter_by_id` to the user defined enum type.
+
+    Args
+    ----
+        parameter_metadata_dict (Dict[int, ParameterMetadata]): Parameter metadata by id.
+
+        parameter_by_id (Dict[int, Any]): Parameters by ID to compare the metadata with.
+
+    Returns
+    -------
+        None: No return value.
+
+    """
+    for i, value in parameter_by_id.items():
+        parameter_metadata = parameter_metadata_dict[i]
+        has_enum_values_annotation, enum_values_annotation = try_get_enum_values_annotation(
+            parameter_metadata
+        )
+        if has_enum_values_annotation:
+            enum_type = _get_enum_type(parameter_metadata)
+            if parameter_metadata.repeated:
+                for j, member_value in enumerate(value):
+                    for enum in enum_type:
+                        if enum.value == member_value:
+                            parameter_by_id[i][j] = enum
+            else:
+                for enum in enum_type:
+                    if enum.value == value:
+                        parameter_by_id[i] = enum
```

Comment:
Iterating over all of the possible enum values is not a performant way to deserialize an enum. How about
`parameter_by_id[i] = enum_type(value)`?

```
>>> from enum import Enum
>>> class Color(Enum):
...     RED = 1
...     GREEN = 2
...     BLUE = 3
...
>>> Color(3)
<Color.BLUE: 3>
```

If the value is not in the enum, your current code leaves it unassigned. Is that really what you want? Or should the measurement service report an error?

Constructing an enum with an unsupported value raises `ValueError`. You can catch this exception and implement whatever behavior you want (leave the value unassigned, set it to None, set it to default, etc.).
```
>>> Color(4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.13\lib\enum.py", line 384, in __call__
    return cls.__new__(cls, value)
  File "C:\Users\Administrator\.pyenv\pyenv-win\versions\3.9.13\lib\enum.py", line 702, in __new__
    raise ve_exc
ValueError: 4 is not a valid Color
```

Reply:
Updated as suggested and to throw TypeError exceptions for when the value is not a valid value within the enum. I thought an error would be more appropriate than leaving it was unassigned, None, or default as if a user is giving us bad values here, we should be letting the user know.